spyder/api/utils.py

```diff
@@ -0,0 +1,1096 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Copyright © <2013-2016> <Colin Duquesnoy and others, see pyqode/AUTHORS.rst>
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+# -*- coding: utf-8 -*-
+"""
+This module contains utility functions/classes.
+Adapted from https://github.com/pyQode/pyqode.core/blob/master/pyqode/core/api/utils.py
+"""
+import functools
+import weakref
+
+from qtpy.QtCore import QTimer, Qt
+from qtpy.QtGui import QColor, QTextCursor, QTextBlock, QTextDocument, QCursor
+
+from qtpy.QtWidgets import QApplication
+
+
+class memoized(object):
```

Comment:
We already have a function in `utils` to do memoizations. Please try to merge this function with that one.